asclepias_broker/harvester/zenodo.py

```diff
@@ -0,0 +1,159 @@
+
+"""Versioning metadata harvester."""
+
+from copy import deepcopy
+from datetime import datetime
+from typing import List
+
+import requests
+from flask import current_app
+
+from .base import MetadataHarvester
+
+
+class ZenodoAPIException(Exception):
+    """Zenodo REST API exception."""
+
+
+class ZenodoClient:
+    """Zenodo client."""
+
+    url = 'https://zenodo.org/api/records/'
+    params = {
+        'page': 1,
+        'size': 100,
+        'all_versions': True,
+        'sort': '-version'
+    }
+
+    def get_concept_doi(self, doi: str) -> str:
+        """."""
+        query = 'doi:"{}"'.format(doi)
+        params = deepcopy(self.params)
+        params['q'] = query
+
+        res = requests.get(self.url, params=params)
+        conceptdoi = None
+        if res.ok:
+            if res.json()['hits']['total'] == 1:
+                conceptdoi = res.json()['hits']['hits'][0].get('conceptdoi')
+            else:
+                conceptdoi = doi  # it's already a conceptdoi
+        else:
+            raise ZenodoAPIException()
+        return conceptdoi
+
+    def get_versions(self, conceptdoi) -> List[str]:
+        """."""
+        query = 'conceptdoi:"{}"'.format(conceptdoi)
+        params = deepcopy(self.params)
+        params['q'] = query
+        url = deepcopy(self.url)
+
+        while True:
+            res = requests.get(url, params=params)
+            if res.ok:
+                data = res.json()
+                for r in data['hits']['hits']:
+                    yield r['doi']
+                if data['links'].get('next'):
+                    url = data['links'].get('next')
+                    params = {}
+                else:
+                    break
+            else:
+                raise ZenodoAPIException()
+
+
+class ZenodoVersioningHarvester(MetadataHarvester):
+    """Metadata harvester for Zenodo records' versioning."""
+
+    def __init__(self, *, provider_name: str = None):
+        """."""
+        self.provider_name = provider_name or "Zenodo versioning harvester"
+
+    def can_harvest(self, identifier: str, scheme: str,
+                    providers: List[str] = None) -> bool:
+        """."""
+        is_provider = False
+        if providers:
+            is_provider = self.provider_name in providers
+
+        return self._is_zenodo_doi(scheme, identifier) and not is_provider
+
+    def harvest(self, identifier: str, scheme: str,
+                providers: List[str] = None):
+        """."""
+        conceptdoi, versions = self.get_versioning_metadata(identifier)
+        if conceptdoi:
+            providers = set(providers) if providers else set()
+            providers.add(self.provider_name)
+            update_versioning(conceptdoi, versions, providers=list(providers))
+
+    def _is_zenodo_doi(self,  scheme: str, identifier: str) -> bool:
+        if scheme.lower() == 'doi' and '10.5281/zenodo.' in identifier:
```

Comment:
`identifier.lower().startswitch('10.5281/')` would be a bit more future-proof and would protect from falsely identifying a DOI like `10.1234/i-make-peoples-lives-difficult-10.5281/zenodo.-because-im-evil` (which is valid).